File: PyRonaScript/rona_lexer/lexer.py
#!/usr/bin/env python
"""
File: rona_lexer.py
Description:
Author: Malcolm Hall
Version: 1

MIT License

Copyright (c) 2020 Malcolm Hall

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""

from rona_token import Token, TokenType, lexeme_sets
from cpp_helpers import *
from .lexer_base import LexerBase

"""

FILE FORMATS:
    - *.rn Source code file
    - *.rnc Compiled byte code file
        * running a byte code file will skip the rona_lexer, rona_parser, and code generation stages, 
            jumping directly to the runtime stage
HANDLERS:
    - string literals
    - bool constants
    - int constants
    - double constants
    - reserved words
    - default handler
    - compound operators
    - comment
    

PARSER NOTES:
    - objects can have an associated type (function, var, etc)
    - scope state will need to have transition functions
    - need scope cleanup function 
    - should have an expect() function to predict the next rona_token
        * raise exception if prediction is wrong
    - Object classification, object table creation, scoping, import files list, 
    
    
BUILT-IN FUNCTIONS:
    - print
    - input
    - range
    - len
    - 

SCOPING:
    - Objects are declared they will be given a scope level  
    - The object store/lookup will go like this:
        scope->name
    - Conflicting names in different scopes will defer to the closest scope
    - When referencing a variable, the scope level will be checked for the variable name
        then, if the variable exists, it will be used.
    - A scope can access resources in all scopes with a lower scope level. Attempting to access
        a higher level will result in an error.
        
    
"""


class Lexer(LexerBase):
    def __init__(self):
        super().__init__()
        self.ignore_comments: bool = True

    def emit(self, token: Token or str, n=1) -> Token:
        """
        Verifies rona_token formatting, pushes onto rona_token rona_stack, and returns the rona_token.

        :param token: rona_token to be emitted.
        :param n: Number of buffer advancements to make.
        :return: rona_token
        """
        self.tok_str = ""
        self.adv_buf(n)

        if isinstance(token, Token):
            self.tokens.append(token)

        else:
            self.tokens.append(Token(token))

        self.tokens[-1].file_pos = self.get_file_pos()
        return self.tokens[-1]

    def is_cmpnd(self) -> bool:
        """
        Checks whether the next sequence of characters represents a compound.

        :return: Boolean indicating whether a compound is present.
        """
        return self.get_cmpnd_candidate() in lexeme_sets.COMPOUNDS

    def get_cmpnd_candidate(self) -> str:
        """
        Gets a string containing the current and next characters in the buffer.

        :return: string
        """
        return self.current() + self.peek()

    def reserved_word_handler(self, lexeme) -> Token:
        """
        Handles reserved word emission.

        :param lexeme: word to be emitted as a rona_token.
        :return: rona_token
        """
        return self.emit(lexeme, 0)

    def default_handler(self, c) -> Token:
        """
        Default emission handler.

        :param c: character
        :return: rona_token
        """
        if self.tok_str == "":
            return self.emit(c)
        else:
            return self.emit(self.tok_str, 0)

    def comment_handler(self, c) -> Token:
        """
        Comment emission handler.

        :param c: character
        :return: rona_token
        """
        if self.tok_str == "":
            if self.peek() == "/":
                while self.current() != "\n":
                    self.tok_str += self.current()
                    self.adv_buf(2)
                return self.emit(Token(self.tok_str, TokenType.COMMENT))
            elif self.peek() == '*':
                idx = 0
                while (self.current() + self.peek()) != "*/" and not self.eof:
                    idx += 1
                    self.tok_str += self.current()
                    self.adv_buf()
                print("BREAK")
                self.tok_str += self.current()
                self.adv_buf()
                self.tok_str += self.current()
                print("HERE")
                return self.emit(Token(self.tok_str, TokenType.COMMENT))
            else:
                 # Throw error
                ...
        else:
            return self.emit(self.tok_str)

    def cmpnd_handler(self, c: str) -> Token:
        """
        Handles compound tokens, other than reserved words or identifiers.

        :param c: character
        :return: rona_token
        """
        if self.tok_str == "":
            if self.is_cmpnd():
                return self.emit(self.get_cmpnd_candidate(), 2)
            else:
                return self.emit(c)
        else:
            return self.emit(self.tok_str, 0)

    def string_literal_handler(self, c: str) -> Token:
        """
        Handles the emission of string literals.

        :param c: 
        :return: rona_token
        """
        if self.tok_str == "":
            self.tok_str += c
            self.adv_buf()

            while self.current() != '"':
                self.tok_str += self.current()
                self.adv_buf()
            self.tok_str += self.current()
            self.adv_buf()

            return self.emit(Token(self.tok_str, TokenType.STRING_LITERAL))
        else:
            return self.emit(self.tok_str)

    def get_token(self) -> Token:
        """
        Main lexing function.

        :return: rona_token
        """

        while not self.eof:
            if self.current() == " ":
                if self.tok_str != "":
                    return self.reserved_word_handler(self.tok_str)
                else:
                    while self.current() == " ":
                        self.adv_buf()

            elif self.current() == "{":
                return self.default_handler(self.current())

            elif self.current() == "}":
                return self.default_handler(self.current())

            elif self.current() == "(":
                return self.default_handler(self.current())

            elif self.current() == ")":
                return self.default_handler(self.current())

            elif self.current() == "[":
                return self.default_handler(self.current())

            elif self.current() == "]":
                return self.default_handler(self.current())

            elif self.current() == "<":
                return self.default_handler(self.current())

            elif self.current() == ">":
                return self.default_handler(self.current())

            elif self.current() == "/":
                if self.peek() == '*':
                    obj = self.comment_handler(self.current())
                    print(self.tokens[-1])
                    print("SOFIJSLDKFJSLF")
                    return obj
                else:
                    return self.cmpnd_handler(self.current())

            elif self.current() == "+":
                return self.cmpnd_handler(self.current())

            elif self.current() == "-":
                return self.cmpnd_handler(self.current())

            elif self.current() == "*":
                return self.cmpnd_handler(self.current())

            elif self.current() == "%":
                return self.cmpnd_handler(self.current())

            elif self.current() == "&":
                return self.cmpnd_handler(self.current())

            elif self.current() == "|":
                return self.cmpnd_handler(self.current())

            elif self.current() == "^":
                return self.default_handler(self.current())

            elif self.current() == "~":
                return self.cmpnd_handler(self.current())

            elif self.current() == "!":
                return self.cmpnd_handler(self.current())

            elif self.current() == "-":
                return self.cmpnd_handler(self.current())

            elif self.current() == "$":
                return self.default_handler(self.current())

            elif self.current() == "/":
                if self.peek() in ['/', '*']:
                    return self.comment_handler(self.current())
                else:
                    return self.cmpnd_handler(self.current())

            elif self.current() == "=":
                return self.cmpnd_handler(self.current())

            elif self.current() == ",":
                return self.default_handler(self.current())

            elif self.current() == ".":
                return self.default_handler(self.current())

            elif self.current() == ":":
                return self.default_handler(self.current())

            elif self.current() == ";":
                return self.default_handler(self.current())

            elif self.current() == '"':
                return self.string_literal_handler(self.current())

            else:
                if self.current() not in ["\r", "\t", "\n"]:
                    self.tok_str += self.current()
                    if self.current().isnumeric():
                        if self.peek() == '.':
                            self.adv_buf()
                            self.tok_str += self.current()
                            self.adv_buf()

                            if self.current().isnumeric():
                                self.tok_str += self.current()

                self.adv_buf()

    def load_tokens(self):
        while not self.eof:
            self.get_token()

        eof_tok = Token(None, TokenType.EOF)
        eof_tok.token = TokenType.EOF
        eof_tok.lexeme = ""

File: src/lexer/Lexer.cpp
/******************************************************************************
* File:
* Description:
* Author: Malcolm Hall
* Version: 1
*
* MIT License
*
* Copyright (c) 2020 - 2023 Malcolm Hall
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
* SOFTWARE.
*******************************************************************************/
#include "Lexer.h"
#include <algorithm>
#include <utility>
#include "Token.h"
#include "../memory_mgmt/RnLinearAllocator.h"

#undef TOKEN_DEF
#undef RESERVED_WORD
#define TOKEN_DEF(token, lexeme) {lexeme, TokenType::token},
#define RESERVED_WORD TOKEN_DEF
std::unordered_map<std::string, TokenType> Lexer::_token_map = {RN_TOKEN_LIST};

#undef TOKEN_DEF
#undef RESERVED_WORD
#define TOKEN_DEF(token, lexeme)
#define RESERVED_WORD(token, lexeme) lexeme,

const std::unordered_set<std::string> Lexer::_reserved_words = {RN_TOKEN_LIST};

const std::unordered_set<std::string> Lexer::_compounds = {
    "++", "+=", "--", "-=", "&&", "%=", "||", "**", "&=", "|=", "/=", "*=", "~=",
    "^=", "!=", "->", "::", ">=", "<=", "==", ">>", "<<", "::", "??", "in"};

const std::unordered_set<std::string> Lexer::_binary_ops = {
    "+"
    "-"
    "=="
    ">="
    "<="
    "!="
    ">"
    "<"
    "&"
    "&&"
    "|"
    "||"
    "^"
    "/"
    "%"
    "*"
    "??"
    "->"
    "::",
    "in"};

/*****************************************************************************/
Lexer::Lexer() {
    file_info.SetFilePath(_file_path);
    FillBuffer('\0');
}

/*****************************************************************************/
Lexer::~Lexer() {
    _file_obj.close();
    for (auto& token : tokens)
        delete token;
}

/*****************************************************************************/
Token* Lexer::Emit(TokenType type) {
    auto token = MakeToken(_lexeme, type);

    // This is a little awkward, but it handles repeated unary operators
    if (token->GetType() == TokenType::NAME &&
        (token->GetLexeme()[0] == '+' || token->GetLexeme()[0] == '-')) {
        tokens.emplace_back(MakeToken(std::string(1, token->GetLexeme()[0])));
        token->SetLexeme(token->GetLexeme().substr(1));
    }
    _lexeme.clear();
    tokens.emplace_back(token);

    return token;
}

/*****************************************************************************/
Token* Lexer::MakeToken(const std::string& s, TokenType initial_type) const {
    auto token = new Token(s, initial_type);
    token->file_info = file_info;

    // A number can have an abitrary number of signs in front of it
    auto normalize_sign = [](const std::string& s) {
        size_t i = 0;
        int sign = 1;
        for (; i < s.length(); i++) {
            if (s[i] == '-') {
                sign *= -1;
                continue;
            } else if (s[i] == '+') {
                continue;
            }
            break;
        }
        if (sign == 1)
            return s.substr(i);
        return std::string(1, '-') + s.substr(i);
    };

    if (initial_type == TokenType::UNDEFINED) {
        if (_token_map.count(s) || (_token_map.count(s) && IsReservedWord(s))) {
            token->SetType(_token_map[s]);
        } else if (IsIntLiteral(s)) {
            token->SetType(TokenType::INT_LITERAL);
            token->SetLexeme(normalize_sign(s));
        } else if (IsFloatLiteral(s)) {
            token->SetType(TokenType::FLOAT_LITERAL);
            token->SetLexeme(normalize_sign(s));
        } else if (IsStrLiteral(s)) {
            token->SetType(TokenType::STRING_LITERAL);
        } else if (IsBoolLiteral(s)) {
            token->SetType(TokenType::BOOL_LITERAL);
        } else {
            token->SetType(TokenType::NAME);
        }
    }
    return token;
}

/*****************************************************************************/
bool Lexer::IsIntLiteral(const std::string& s) {
    size_t i = 0;
    char c;
    for (; i < s.length(); i++) {
        c = s[i];
        if (c != '+' && c != '-')
            break;
    }
    for (; i < s.length(); i++) {
        c = s[i];
        if (!std::isdigit(c))
            return false;
    }
    return true;
}

/*****************************************************************************/
bool Lexer::IsFloatLiteral(const std::string& s) {
    bool decimal_found = false;
    size_t i = 0;
    char c;
    for (; i < s.length(); i++) {
        c = s[i];
        if (c != '+' && c != '-')
            break;
    }

    for (; i < s.length(); ++i) {
        c = s[i];
        if (!std::isdigit(c)) {
            if (c == '.' && !decimal_found)
                decimal_found = true;
            else
                return false;
        }
    }
    return decimal_found;
}

/*****************************************************************************/
bool Lexer::IsHexLiteral(const std::string& s) {
    std::string prefix = std::string(1, s[0]) + std::string(1, s[1]);
    if (prefix == "0x" || prefix == "0X") {
        for (size_t i = 2; i < s.length(); ++i) {
            if (!std::isdigit(s[i]))
                return false;
        }
        return true;
    } else {
        return false;
    }
}

/*****************************************************************************/
bool Lexer::IsBoolLiteral(const std::string& s) {
    return s == "true" || s == "false";
}

/*****************************************************************************/
bool Lexer::IsStrLiteral(const std::string& s) {
    return (s[0] == '"' && s[s.length() - 1] == '"') ||
           (s[0] == '\'' && s[s.length() - 1] == '\'');
}

/*****************************************************************************/
bool Lexer::IsCompound() const {
    if (Peek() == '>') {
        return true;
    }
    if ((Current() == '-' || Current() == '+') &&
        (tokens.back()->IsLiteral() ||
         _binary_ops.find(std::string(1, Peek())) == _binary_ops.end()) &&
        Peek() != '=') {
        return false;
    }
    return _compounds.find(GetCompoundCandidate()) != _compounds.end();
}

/*****************************************************************************/
bool Lexer::IsReservedWord(const std::string& s) {
    return _reserved_words.find(s) != _reserved_words.end();
}

/*****************************************************************************/
std::string Lexer::GetCompoundCandidate(int n) const {
    std::string s = std::string(1, Current());

    for (int i = 1; i < n; ++i)
        s.append(std::string(1, Peek(i)));

    return s;
}

/*****************************************************************************/
Token* Lexer::ProcessReservedWord() {
    return Emit();
}

/*****************************************************************************/
Token* Lexer::ProcessDefault() {
    if (_lexeme.empty()) {
        _lexeme = std::string(1, Current());
        AdvanceBuffer(1);
    }
    return Emit();
}

/*****************************************************************************/
Token* Lexer::ProcessComment(bool is_block_comment) {
    if (_lexeme.empty()) {
        if (is_block_comment) {
            while (GetCompoundCandidate() != BLOCK_COMMENT_END && !EndOfSequence())
                AdvanceBuffer(1);
        } else {
            while (Current() != INLINE_COMMENT_END_N &&
                   Current() != INLINE_COMMENT_END_R && !EndOfSequence())
                AdvanceBuffer(1);
        }
        AdvanceBuffer(1);

        if (is_block_comment)
            AdvanceBuffer(1);
        return new Token("", TokenType::UNDEFINED);
    } else {
        return Emit();
    }
}

/*****************************************************************************/
Token* Lexer::ProcessOperator() {
    if (_lexeme.empty()) {
        if (IsCompound()) {
            _lexeme = GetCompoundCandidate();
            AdvanceBuffer(2);
        } else {
            _lexeme = std::string(1, Current());
            AdvanceBuffer(1);
        }
    }

    return Emit();
}

/*****************************************************************************/
Token* Lexer::ProcessStringLiteral() {
    if (_lexeme.empty()) {
        AdvanceBuffer(1);

        while (_lexeme.length() < STRING_LITERAL_MAX_LENGTH) {
            if (Current() == '\"' && Lookback() != '\\') {
                AdvanceBuffer(1);
                break;
            } else if (Current() == '\\') {
                switch (Peek()) {
                    case 'n':
                        _lexeme += '\n';
                        AdvanceBuffer(1);
                        break;
                    case 't':
                        _lexeme += '\t';
                        AdvanceBuffer(1);
                        break;
                    case 'r':
                        _lexeme += '\r';
                        AdvanceBuffer(1);
                        break;
                    default:
                        _lexeme += Current();
                        break;
                }
            } else {
                _lexeme += Current();
            }
            AdvanceBuffer(1);
        }
    }

    return Emit(TokenType::STRING_LITERAL);
}

/*****************************************************************************/
Token* Lexer::Consume() {
    switch (Current()) {
        case '\n':
        case '\t':
        case '\r':
        case ' ':
            if (!_lexeme.empty()) {
                AdvanceBuffer(1);
                return ProcessReservedWord();
            } else {
                while (IsWhiteSpace(Current()) && !EndOfSequence())
                    AdvanceBuffer(1);
                break;
            }
        case '{':
        case '}':
        case '[':
        case ']':
        case '(':
        case ')':
        case ',':
        case ';':
            return ProcessDefault();
        case '/':
            if (GetCompoundCandidate() == BLOCK_COMMENT_START)
                return ProcessComment(true);
            else if (GetCompoundCandidate() == INLINE_COMMENT_START)
                return ProcessComment(false);
            else
                return ProcessOperator();
        case '*':
        case '^':
        case '%':
        case '&':
        case '|':
        case '~':
        case '=':
        case '<':
        case '>':
        case '!':
        case ':':
            return ProcessOperator();
        case '\'':
        case '\"':
            return ProcessStringLiteral();
        case '+':
        case '-': {
            if (Peek() == Current() && tokens.back()->IsOneOf({TokenType::NAME})) {
                if (!_lexeme.empty()) {
                    Emit();
                }
                _lexeme = GetCompoundCandidate();
                AdvanceBuffer(2);
                return Emit();
            } else {
                if ((tokens.back()->GetType() != TokenType::R_PARAN &&
                     !tokens.back()->IsBinaryOp()) ||
                    !_lexeme.empty() || Peek() == '=' || Peek() == '>')
                    return ProcessOperator();
            }
        }
        default: {
            if (Current() != '\r' && Current() != '\t' && Current() != '\n')
                _lexeme += Current();
            AdvanceBuffer(1);
            break;
        }
    }
    return new Token("", TokenType::UNDEFINED);
}

/*****************************************************************************/
void Lexer::ProcessTokens() {
    while (!EndOfSequence())
        Consume();
    Consume();

    // Clear the last _lexeme
    if (!_lexeme.empty())
        Emit();
}

/*****************************************************************************/
void Lexer::LoadFile(const std::string& path) {
    file_info.SetFilePath(path);
    _file_obj.open(file_info.GetFilePath());

    if (_file_obj.fail())
        throw std::runtime_error("Failed to open " + file_info.GetFilePath());

    auto file_size = std::filesystem::file_size(path);
    _data = new char[file_size];
    std::memset(_data, 0, file_size);
    _file_obj.read(_data, file_size);
    _data_size = file_size;
    _file_obj.close();
    AdvanceBuffer(2);
}

/*****************************************************************************/
char Lexer::GetCurrentAsExpectedType() {
    return Current();
}

/*****************************************************************************/
void Lexer::RunAdvanceBufferSideEffects() {
    file_info.IncrementCharNum();

    if (Lookback() == '\n' || Lookback() == '\r') {
        file_info.IncrementLineNum();
        file_info.SetCharNum(1);
        file_info.UpdateLineStartValues(_current_line_start);
        _current_line_start = _data_idx;
    }
}

/*****************************************************************************/
bool Lexer::IsWhiteSpace(char c) {
    std::vector<char> ws_chars = {'\r', '\t', ' ', '\n'};
    return std::find(ws_chars.begin(), ws_chars.end(), c) != ws_chars.end();
}

/*****************************************************************************/
std::string Lexer::ItemToString(char item) {
    return {1, item};
}

/*****************************************************************************/
void Lexer::Reset() {
    for (auto token : tokens) {
        delete token;
    }
    tokens.clear();
}

/*****************************************************************************/
void Lexer::Run() {
    ProcessTokens();
}

/*****************************************************************************/
void Lexer::HandleUnexpectedItem() {
    std::string msg = "Unexpected character '" + std::string(1, Current()) + "'" +
                      " in " + file_info.ToString() + "\nExpected one of [";

    for (auto& expected_char : _expected_items)
        msg.append("'" + std::string(1, expected_char) + "', ");

    if (!_expected_items.empty())
        msg = msg.substr(0, msg.length() - 2);

    msg += "]\n\n";
    msg += file_info.GetContextualBlock();

    throw std::runtime_error(msg);
}
